# Coding Exercises (Part 3)

Now, you will have the opportunity to analyze your own dataset. <br>
__Follow the instructions__ and insert your code! You are either requested to 
- Complete the Code and __Fill in the gaps__. Gaps are marked with "__---__" and are __placeholders__ for your code fragment. 
- Write Code completely __on your own__ 

In some exercises, you will find questions that can only be answered, if your code is correct and returns the right output! The correct answer is provided below your coding cell. There you can check whether your code is correct.

If you need a hint, check the __Hints Section__ at the end of this Notebook. Exercises and Hints are numerated accordingly.

If you need some further help or if you want to check your code, you can also watch the __solutions videos__ or check the __solutions notebook__.

### Have Fun!

--------------------------------------------------------------------------------------------------------------

## DataFrame Basics II

### Exercise 5: Filtering DataFrames

In [1]:
#run the cell!
import pandas as pd

In [2]:
#run the cell!
cars = pd.read_csv("cars.csv")

In [3]:
#run the cell!
cars.head()

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0  18.0          8         307.0       130.0    3504          12.0   
1  15.0          8         350.0       165.0    3693          11.5   
2  18.0          8         318.0       150.0    3436          11.0   
3  16.0          8         304.0       150.0    3433          12.0   
4  17.0          8         302.0       140.0    3449          10.5   

   model_year origin                       name  
0          70    usa  chevrolet chevelle malibu  
1          70    usa          buick skylark 320  
2          70    usa         plymouth satellite  
3          70    usa              amc rebel sst  
4          70    usa                ford torino

40. __Check__ for all elements in the column __"origin"__ whether they are __equal to "europe"__. __Save__ the boolean Series in the variable __mask1__! __Fill in the gaps__!

In [4]:
mask1 = cars.origin == "europe"
mask1

0      False
1      False
2      False
3      False
4      False
5      False
6      False
7      False
8      False
9      False
10     False
11     False
12     False
13     False
14     False
15     False
16     False
17     False
18     False
19      True
20      True
21      True
22      True
23      True
24     False
25     False
26     False
27     False
28     False
29     False
       ...  
368    False
369    False
370    False
371    False
372    False
373    False
374    False
375     True
376    False
377    False
378    False
379    False
380    False
381    False
382    False
383    False
384    False
385    False
386    False
387    False
388    False
389    False
390    False
391    False
392    False
393    False
394     True
395    False
396    False
397    False
Name: origin, Length: 398, dtype: bool

41. __Check__ for all elements in the column __"mpg"__ whether they are __smaller than 20__. __Save__ the boolean Series in the variable __mask2__!

In [5]:
mask2 = cars.mpg < 20

42. __Filter__ the cars DataFrame for __all cars from europe__ (use mask1) and __save__ the subset in the variable __europe__! __Fill in the gaps!__

In [6]:
europe  = cars.loc[mask1].copy()

__Inspect__! What is the name of the __second__ car?

In [7]:
# run the cell!
europe.head()

mpg  cylinders  displacement  horsepower  weight  acceleration  \
19  26.0          4          97.0        46.0    1835          20.5   
20  25.0          4         110.0        87.0    2672          17.5   
21  24.0          4         107.0        90.0    2430          14.5   
22  25.0          4         104.0        95.0    2375          17.5   
23  26.0          4         121.0       113.0    2234          12.5   

    model_year  origin                  name  
19          70  europe  vw 1131 deluxe sedan  
20          70  europe           peugeot 504  
21          70  europe           audi 100 ls  
22          70  europe              saab 99e  
23          70  europe              bmw 2002

The second car is a ... peugeot 504!

43. Get some __meta information__ on the DataFrame europe. __How many__ cars are from europe? 

In [8]:
europe.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 70 entries, 19 to 394
Data columns (total 9 columns):
mpg             70 non-null float64
cylinders       70 non-null int64
displacement    70 non-null float64
horsepower      68 non-null float64
weight          70 non-null int64
acceleration    70 non-null float64
model_year      70 non-null int64
origin          70 non-null object
name            70 non-null object
dtypes: float64(4), int64(3), object(2)
memory usage: 5.5+ KB


... 70 cars are from europe!

44. __Filter__ the DataFrame __cars__ for all cars __from europe with low fuel efficiency__ (mpg lower than 20). Use __mask1__ and __mask2__!<br> __Save__ the subset in the variable __europe_le__! __Fill in the gaps__!

In [9]:
europe_le = cars.loc[mask1 & mask2].copy()

__Inspect__! What is the __name__ of the __least efficient__ car from europe? 

In [10]:
#run the cell!
europe_le

mpg  cylinders  displacement  horsepower  weight  acceleration  \
76   18.0          4         121.0       112.0    2933          14.5   
120  19.0          4         121.0       112.0    2868          15.5   
209  19.0          4         120.0        88.0    3270          21.9   
211  16.5          6         168.0       120.0    3820          16.7   
275  17.0          6         163.0       125.0    3140          13.6   
277  16.2          6         163.0       133.0    3410          15.8   

     model_year  origin                name  
76           72  europe     volvo 145e (sw)  
120          73  europe         volvo 144ea  
209          76  europe         peugeot 504  
211          76  europe  mercedes-benz 280s  
275          78  europe         volvo 264gl  
277          78  europe       peugeot 604sl

The least efficient european car is ... peugeot 604sl

45. __Filter__ the DataFrame cars for all cars with an __mpg between 10 and 15__ (both ends inclusive!). <br>
 __Save__ the subset in the variable __mpg_10_15__! __Fill in the gaps__!

In [11]:
mpg_10_15 = cars.loc[cars.mpg.between(10,15)].copy()

__Inspect__! The __first__ car is...?

In [12]:
# run the cell!
mpg_10_15.head()

mpg  cylinders  displacement  horsepower  weight  acceleration  \
1  15.0          8         350.0       165.0    3693          11.5   
5  15.0          8         429.0       198.0    4341          10.0   
6  14.0          8         454.0       220.0    4354           9.0   
7  14.0          8         440.0       215.0    4312           8.5   
8  14.0          8         455.0       225.0    4425          10.0   

   model_year origin               name  
1          70    usa  buick skylark 320  
5          70    usa   ford galaxie 500  
6          70    usa   chevrolet impala  
7          70    usa  plymouth fury iii  
8          70    usa   pontiac catalina

The __first__ car is... buick skylark 320!

__Inspect__! __How many__ cars are in the subset?

In [13]:
#run the cell!
mpg_10_15.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 68 entries, 1 to 224
Data columns (total 9 columns):
mpg             68 non-null float64
cylinders       68 non-null int64
displacement    68 non-null float64
horsepower      68 non-null float64
weight          68 non-null int64
acceleration    68 non-null float64
model_year      68 non-null int64
origin          68 non-null object
name            68 non-null object
dtypes: float64(4), int64(3), object(2)
memory usage: 5.3+ KB


There are ... 68 cars!

46. __Filter__ the Dataframe __cars__ for all cars that are __not built in the years 73 and 74__, and only the columns __"mpg"__ and __"name"__!<br> __Save__ the subset in the variable __not_73_74__. __Fill in the gaps!__ 

In [13]:
not_73_74 = cars.loc[~cars.model_year.isin([73, 74]), ["mpg", "name"]].copy()

In [14]:
#run the cell!
not_73_74.head()

mpg                       name
0  18.0  chevrolet chevelle malibu
1  15.0          buick skylark 320
2  18.0         plymouth satellite
3  16.0              amc rebel sst
4  17.0                ford torino

__Inspect__! __How many__ cars are in this subset?

In [15]:
# run the cell!
not_73_74.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 331 entries, 0 to 397
Data columns (total 2 columns):
mpg     331 non-null float64
name    331 non-null object
dtypes: float64(1), object(1)
memory usage: 7.8+ KB


There are ... 331 cars in the subset! 

In [16]:
# run the cell!
cars.head()

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0  18.0          8         307.0       130.0    3504          12.0   
1  15.0          8         350.0       165.0    3693          11.5   
2  18.0          8         318.0       150.0    3436          11.0   
3  16.0          8         304.0       150.0    3433          12.0   
4  17.0          8         302.0       140.0    3449          10.5   

   model_year origin                       name  
0          70    usa  chevrolet chevelle malibu  
1          70    usa          buick skylark 320  
2          70    usa         plymouth satellite  
3          70    usa              amc rebel sst  
4          70    usa                ford torino

47. __Drop__ the columns __"displacement"__ and __"cylinders"__ from the DataFrame cars! __Save__ the change!

In [17]:
cars.drop(columns = ["displacement", "cylinders"], inplace = True)

In [18]:
# run the cell!
cars.head()

mpg  horsepower  weight  acceleration  model_year origin  \
0  18.0       130.0    3504          12.0          70    usa   
1  15.0       165.0    3693          11.5          70    usa   
2  18.0       150.0    3436          11.0          70    usa   
3  16.0       150.0    3433          12.0          70    usa   
4  17.0       140.0    3449          10.5          70    usa   

                        name  
0  chevrolet chevelle malibu  
1          buick skylark 320  
2         plymouth satellite  
3              amc rebel sst  
4                ford torino

48. __Drop__ all rows/cars from __"usa"__ from the DataFrame cars! Use the __loc__ operator and __overwrite cars__!

In [19]:
cars = cars.loc[cars.origin != "usa"]

In [20]:
# run the cell!
cars.head()

mpg  horsepower  weight  acceleration  model_year  origin  \
14  24.0        95.0    2372          15.0          70   japan   
18  27.0        88.0    2130          14.5          70   japan   
19  26.0        46.0    1835          20.5          70  europe   
20  25.0        87.0    2672          17.5          70  europe   
21  24.0        90.0    2430          14.5          70  europe   

                     name  
14  toyota corona mark ii  
18           datsun pl510  
19   vw 1131 deluxe sedan  
20            peugeot 504  
21            audi 100 ls

49. __Add__ the new column __"l_per_100km"__ to the DataFrame cars by calculating __235.21/mpg__. __Round__ to 2 decimals. __Fill in the gaps__!

In [21]:
cars["l_per_100km"] = (235.21/cars.mpg).round(2)

__Inspect__ cars! What is the __"l_per_100km"__ value for the __audi 100 ls__?

In [22]:
# run the cell!
cars.head()

mpg  horsepower  weight  acceleration  model_year  origin  \
14  24.0        95.0    2372          15.0          70   japan   
18  27.0        88.0    2130          14.5          70   japan   
19  26.0        46.0    1835          20.5          70  europe   
20  25.0        87.0    2672          17.5          70  europe   
21  24.0        90.0    2430          14.5          70  europe   

                     name  l_per_100km  
14  toyota corona mark ii         9.80  
18           datsun pl510         8.71  
19   vw 1131 deluxe sedan         9.05  
20            peugeot 504         9.41  
21            audi 100 ls         9.80

The "l_per_100km" value for the audi 100 ls is ... 9.80!

# Well Done!

---------------------------------

# Hints (Spoiler!)

40. origin attribute; "europe"

41. mask2  =  cars.---  <  --- 

42. loc operator, mask1

43. info() method

44. Condition1 & Condition2

45. between() method

46. isin() method and ~ (tilde symbol)

47. drop() method

48. use != (not equal)

49. cars["new_column_label"]